rust/kernel/clk.rs

```diff
@@ -0,0 +1,54 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Common clock framework.
+//!
+//! C header: [`include/linux/clk.h`](../../../../include/linux/clk.h)
+
+use crate::{bindings, device, error::Result, str::CStr, to_result};
+
+/// Represents `struct clk *`.
+///
+/// # Invariants
+///
+/// The pointer is valid.
+pub struct Clk(*mut bindings::clk);
+
+impl Clk {
+    /// Creates new clock structure from a raw pointer.
+    ///
+    /// # Safety
+    ///
+    /// The pointer must be valid.
+    pub unsafe fn new(clk: *mut bindings::clk) -> Self {
+        Self(clk)
+    }
+
+    /// Lookups and obtains a managed reference to a clock producer.
+    ///
+    /// * `dev` - device for clock "consumer".
+    /// * `id` - ID of clock consumer.
+    pub fn get(dev: &impl device::RawDevice, id: Option<&CStr>) -> Result<Self> {
+        dev.clk_get(id)
+    }
+
+    /// Returns value of the rate field of `struct clk`.
+    pub fn get_rate(&self) -> usize {
+        // SAFETY: the pointer is valid by the type invariant.
+        unsafe { bindings::clk_get_rate(self.0) as usize }
+    }
+
+    /// Prepares and enables the underlying hardware clock.
+    ///
+    /// This function should not be called in atomic context.
```

Comment:
If so than this is an unsafe fn.

Reply:
Basically this is what `unsafe` for - to tell you there is something you should pay attention to, otherwise things may blow up.

Reply:
> Basically this is what `unsafe` for - to tell you there is something you should pay attention to, otherwise things may blow up.

Thanks for your attempt at explaining `unsafe`, @kpp, it is appreciated.

You (perhaps purposefully?) leave out the meaning of 'blow up' though. Does it include deadlocks? If so, we need to make _all_ lock acquisition unsafe. Does it include panicking? If so, we need to make accessing array/slice elements by index unsafe because we may panic if the index is out of bounds; we'll also need to make _all_ functions that may result in infallible allocations unsafe as well (e.g., `Box::new`) because they may panic if the allocation fails. Of course these are not included.

Anyway, `unsafe` is used to delimit expressions/statements whose freedom of undefined behaviour depends on facts that cannot be enforced/checked automatically by the compiler. That's it. What I am saying is that calling a sleepable function from atomic context is wrong and shouldn't happen, but it doesn't violate memory safety. Therefore it doesn't need to be marked with `unsafe`. (Very much like the cases exemplified in the previous paragraph.)

We are, of course, aware of it. We have even brought it up with the Rust compiler folks: there's an opportunity here for us to catch possible violations _at compile time_. If this is something that you might be interested in working on, let me know and we talk further.

Reply:
> What I am saying is that calling a sleepable function from atomic context is wrong and shouldn't happen, but it doesn't violate memory safety.

Thank you! I was not aware of that. Now it makes sense.

> you might be interested in working on

Yes

Reply:
@kpp @wedsonaf 
Thanks for the discussion.
As it seems the issue has been addressed, I will mark it as resolved.